cli/lib/png2src.js

```diff
@@ -114,76 +114,82 @@ const ALIASES = {
     cpp: "c",
 };
 
-function run (sourceFile) {
+function run(sourceFile) {
     const png = pngjs.PNG.sync.read(fs.readFileSync(sourceFile), {
         colorType: 1,
         inputColorType: 1,
     });
 
-    if (!png.palette) {
-        throw new Error("Does not have indexed color palette");
-    }
-
-    // Map rgba to palette index
-    const palette = new Map();
-    for (let ii = 0; ii < png.palette.length; ++ii) {
-        const rgba = png.palette[ii];
-        const packed = (rgba[0] << 24) | (rgba[1] << 16) | (rgba[2] << 8) | rgba[3];
-        palette.set(packed, ii);
+    const palette = [];
+    for (let y = 0; y < png.height; ++y) {
+        for (let x = 0; x < png.width; ++x) {
+            const idx = 4 * (png.width * y + x);
+            const r = png.data[idx];
+            const g = png.data[idx + 1];
+            const b = png.data[idx + 2];
+            const a = png.data[idx + 3];
+            const packed = (r << 24) | (g << 16) | (b << 8) | a;
+            let s = palette.findIndex(v => v === packed);
+            if (s === -1) {
+                palette.push(packed);
+            }
+        }
     }
+    palette.sort();
+    console.log(palette);
 
     let flags, flagsHumanReadable, odinFlags;
     let bpp;
-    if (palette.size <= 2) {
+    if (palette.length <= 2) {
         bpp = 1;
         flags = 0;
         flagsHumanReadable = "BLIT_1BPP";
         odinFlags = "nil"
-    } else if (palette.size <= 4) {
+    } else if (palette.length <= 4) {
         bpp = 2;
         flags = 1;
         flagsHumanReadable = "BLIT_2BPP";
         odinFlags = "{ .USE_2BPP }"
     } else {
-        throw new Error("Palette is larger than 4 colors");
+        throw new Error(`Palette has ${palette.length} colors. But you can only have 4 colors.`);
     }
 
     const factor = 8 / bpp;
     if (png.width % factor != 0) {
         throw new Error(`${bpp}BPP sprites must have a width divisible by ${factor}`);
     }
 
-    const bytes = new Uint8Array(png.width*png.height*bpp/8);
+    const bytes = new Uint8Array(png.width * png.height * bpp / 8);
 
     // Read a color (palette index) from the source png
-    function readColor (x, y) {
-        const idx = 4*(png.width*y + x);
+    function readColor(x, y) {
+        const idx = 4 * (png.width * y + x);
         const r = png.data[idx];
-        const g = png.data[idx+1];
-        const b = png.data[idx+2];
-        const a = png.data[idx+3];
+        const g = png.data[idx + 1];
+        const b = png.data[idx + 2];
+        const a = png.data[idx + 3];
         const packed = (r << 24) | (g << 16) | (b << 8) | a;
-        return palette.get(packed);
+        return palette.findIndex(v => v == packed);
```

Comment:
Doing a `findIndex` on every pixel is kinda slow. It would be better to keep a Map of RGBA color to palette index.

Reply:
While I don't think this will affect the overall performance (you don't call this 60 times per second on big images), I'll change it nonetheless.